database/migrations/2021_08_06_184400_refactor_album_model.php

```diff
@@ -0,0 +1,823 @@
+			$table->unsignedBigInteger($name . '_id')->nullable(false);
+			// Indices and constraint definitions
+			$table->foreign('user_id')->references('id')->on('users')->cascadeOnUpdate()->cascadeOnDelete();
+			$table->foreign($name . '_id')->references('id')->on($name . 's')->cascadeOnUpdate()->cascadeOnDelete();
+		});
+	}
+
+	/**
+	 * Creates the table `photos` acc. to the new architecture.
+	 */
+	private function createPhotoTableUp(): void
+	{
+		Schema::create('photos', function (Blueprint $table) {
+			// Column definitions
+			$table->bigIncrements('id')->nullable(false);
+			$table->dateTime('created_at')->nullable(false);
+			$table->dateTime('updated_at')->nullable(false);
+			$table->unsignedInteger('owner_id')->unsinged()->nullable(false)->default(0);
+			$table->unsignedBigInteger('album_id')->nullable()->default(null);
+			$table->string('title', 100)->nullable(false);
+			$table->text('description')->nullable();
+			$table->text('tags')->nullable();
+			$table->string('license', 20)->nullable(false)->default('none');
+			$table->boolean('is_public')->nullable(false)->default(false);
+			$table->boolean('is_starred')->nullable(false)->default(false);
+			$table->string('iso')->nullable()->default(null);
+			$table->string('make')->nullable()->default(null);
+			$table->string('model')->nullable()->default(null);
+			$table->string('lens')->nullable()->default(null);
+			$table->string('aperture')->nullable()->default(null);
+			$table->string('shutter')->nullable()->default(null);
+			$table->string('focal')->nullable()->default(null);
+			$table->decimal('latitude', 10, 8)->nullable()->default(null);
+			$table->decimal('longitude', 11, 8)->nullable()->default(null);
+			$table->decimal('altitude', 10, 4)->nullable()->default(null);
+			$table->decimal('img_direction', 10, 4)->nullable()->default(null);
+			$table->string('location')->nullable()->default(null);
+			$table->dateTime('taken_at')->nullable(true)->default(null)->comment('relative to UTC');
+			$table->string('taken_at_orig_tz', 31)->nullable(true)->default(null)->comment('the timezone at which the photo has originally been taken');
+			$table->string('type', 30)->nullable(false);
+			$table->unsignedBigInteger('filesize')->nullable(false)->default(0);
+			$table->string('checksum', 40)->nullable(false);
+			$table->string('live_photo_short_path')->nullable()->default(null);
+			$table->string('live_photo_content_id')->nullable()->default(null);
+			$table->string('live_photo_checksum', 40)->nullable()->default(null);
+			// Indices and constraint definitions
+			$table->foreign('owner_id')->references('id')->on('users');
+			$table->foreign('album_id')->references('id')->on('albums');
+			$table->index('created_at');
+			$table->index('updated_at');
+			$table->index('taken_at');
+			$table->index('checksum');
+			$table->index('live_photo_content_id');
+			$table->index('live_photo_checksum');
+			$table->index('is_public');
+			$table->index('is_starred');
```

Comment:
> So all these indices are new, correct?

Yes.

> Do they provide a noticeable speedup?

I haven't tested that yet. If was more of a gut feeling that these indices might be helpful based on the queries I had in the back of my mind. Moreover, my installation is not large enough to notice any difference (just around 2,000 photos). However, as I found out PostgreSQL actually needs combined, multi-column indices which cover all the columns of a `WHERE` and `ORDER BY` clause of a query for the most benefit, i.e. if we have a query like `SELECT * FROM photos WHERE photos.owner_id = ? OR photos.is_public = true ORDER BY is_starred DESC, created_at DESC` then a multi-column index which covers `owner_id`, `is_public`, `is_starred` and `created_at` together is required.

So maybe, it is best to drop these indices again.

> And I'm guessing that we don't have an index for title/description because of natural sorting?

That a reason, too, but not what I was thinking of.

I was thinking about columns which are part of an equality check in a `WHERE` clause. I was not thinking about ordering. Moreover, indices cannot be used for `WHERE` clauses with the `LIKE` operator. So I did not create indices for `title` and `description`.

Also note, that an index cannot be used to speed up `ORDER BY`, if the index is over the whole set, but only a subset is selected due to a `WHERE` clause. As we nearly always have a `WHERE` clause (do we ever query for all photos?), I did not create indices for ordering. If you want to speed up both, i.e. filtering and ordering, we need specific multi-column indices which match the use case (see above).

> Is there no value in adding them for albums as well?

Good questions which i unable to answer without thorough research. At the moment I am considering to remove all indices again, because single-column indices probably won't help very much. Defining useful multi-column indices is a complicated task. For PostgreSQL best approach is to use `EXPLAIN ANALYZE` for every real-world SQL query, investigate the compiled query plan and then add the correct combination of indices.


Long story short: I don't know what to do. :laughing: